_drafts/2023-05-11-ota-for-embedded-linux-devices.md

```diff
@@ -0,0 +1,446 @@
+On some other systems, the environment may be stored at a fixed location in
+Flash, or on an EEPROM.
+
+Second, note that the boot script helps with three essential components:
+
+- It loads the kernel from the new root partition. This means we can update the
+  kernel, and it’s important to remark here that it works because the kernel is
+  part of the system image instead of being in a `/boot` partition ;
+- It tells the kernel which system partition to mount as root:
+  `root=/dev/mmcblk0p${rootpart}` ;
+- Finally, it loads the device tree and passes it to the kernel (with the
+  `fdt addr ${fdt_addr}` command). This is highly specific to the Raspberry Pi
+  so we will refer the interested reader
+  [to this article](https://dius.com.au/2015/08/19/raspberry-pi-u-boot/) for
+  more details. Again, this is something that is specific to your board and
+  should be received with your board support package.
+
+## Notifying the server that the update has been received
+
+Our device reboots with the new kernel and the new filesystem. The final step of
+our update cycle is to notify the update server that the update has been
+installed successfully.
+
+Looking once again at `/usr/lib/swupdate/conf.d/09-swupdate-args`, we notice the
+following lines:
+
+```bash
+state=$(fw_printenv ustate | cut -f 2 -d'=')
+if [ "$state" == 1 ]; then
+ # Confirm update Success to server
+        SWUPDATE_SURICATTA_ARGS="-c 2"
+else
+        SWUPDATE_SURICATTA_ARGS=" "
+fi
+```
+
+The first interesting command here is `fw_printenv` which allows us to read the
+U-Boot environment file from our system. For example, you can run
+`fw_printenv rootpart` to see on which partition you are currently running.
+
+The second interesting element here is the variable `ustate` that is read by
+`swupdate`. This variable was set to `1` by `swupdate` after installing the
+update. It is used here to notify the update server that we have successfully
+installed the update. It can also be used by more complex bootloader scripts to
+identify that an image is “under test” and can be reverted if it does not boot
+well.
+
+## Conclusion
+
+We have walked through all the steps required to perform OTA on an embedded
+Linux device. All the mysteries and tricks have been revealed. Following the
+strong principles of the Unix world, we use simple parts and basic communication
+primitives to build a much more complex system.
+
+Knowing now what are the parts and the channels you are equipped to debug issues
+and tackle some additional challenges like delta updates and secure boot.
+
+Let us know in the comment below what else you would like to read about in the
+world of embedded Linux devices!
+
+# Further Reading
+
+- The excellent swupdate manual
+  [https://sbabic.github.io/swupdate/swupdate.html](https://sbabic.github.io/swupdate/swupdate.html)
```

Comment:
if u are lazy like me, you can also do such as `<https://sbabic.github.io/swupdate/swupdate.html>` and get the same result 🥹

Reply:
Or copy/paste from Notion which did it for me 😝